src/Libraries/Microsoft.Extensions.AI.Abstractions/Utilities/AIJsonSchemaTransformCache.cs

```diff
@@ -0,0 +1,79 @@
+/// implementations that enforce vendor-specific restrictions on what constitutes a valid JSON schema for a given function or response format.
+/// </para>
+/// <para>
+/// It is recommended <see cref="IChatClient"/> implementations with schema transformation requirements should create a single static instance of this cache.
+/// </para>
+/// </remarks>
+public sealed class AIJsonSchemaTransformCache
+{
+    private readonly ConditionalWeakTable<AIFunction, object> _functionSchemaCache = new();
+    private readonly ConditionalWeakTable<ChatResponseFormatJson, object> _responseFormatCache = new();
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AIJsonSchemaTransformCache"/> class with the specified options.
+    /// </summary>
+    /// <param name="transformOptions">The options governing schema transformation.</param>
+    public AIJsonSchemaTransformCache(AIJsonSchemaTransformOptions transformOptions)
+    {
+        _ = Throw.IfNull(transformOptions);
+
+        if (transformOptions == AIJsonSchemaTransformOptions.Default)
+        {
+            throw new ArgumentException("The options instance does not specify any transformations.", nameof(transformOptions));
+        }
+
+        TransformOptions = transformOptions;
+    }
+
+    /// <summary>
+    /// Gets the options governing schema transformation.
+    /// </summary>
+    public AIJsonSchemaTransformOptions TransformOptions { get; }
+
+    /// <summary>
+    /// Gets or creates a transformed JSON schema for the specified <see cref="AIFunction"/> instance.
+    /// </summary>
+    /// <param name="function">The function whose JSON schema we want to transform.</param>
+    /// <returns>The transformed JSON schema corresponding to <see cref="TransformOptions"/>.</returns>
+    public JsonElement GetOrCreateTransformedSchema(AIFunction function)
+    {
+        _ = Throw.IfNull(function);
+        return (JsonElement)_functionSchemaCache.GetValue(function, function => AIJsonUtilities.TransformSchema(function.JsonSchema, TransformOptions));
+    }
+
+    /// <summary>
+    /// Gets or creates a transformed JSON schema for the specified <see cref="ChatResponseFormatJson"/> instance.
+    /// </summary>
+    /// <param name="responseFormat">The function whose JSON schema we want to transform.</param>
+    /// <returns>The transformed JSON schema corresponding to <see cref="TransformOptions"/>.</returns>
+    public JsonElement? GetOrCreateTransformedSchema(ChatResponseFormatJson responseFormat)
+    {
+        _ = Throw.IfNull(responseFormat);
+
+        if (responseFormat.Schema is null)
+        {
+            return null;
+        }
+
+        return (JsonElement)_responseFormatCache.GetValue(responseFormat, responseFormat => AIJsonUtilities.TransformSchema(responseFormat.Schema!.Value, TransformOptions));
```

Comment:
Nit: just because I had to stare at it to prove to myself it wasn't closing over anything
```suggestion
        return (JsonElement)_responseFormatCache.GetValue(responseFormat, static responseFormat => AIJsonUtilities.TransformSchema(responseFormat.Schema!.Value, TransformOptions));
```

Reply:
It's closing over `this` for the `TransformOptions` property.